File: packages/uns-core/src/tools/auth/secure-store.ts
import os from "os";
import path from "path";
import fs from "fs";

/**
 * Abstraction over secure storage for tokens.
 *
 * Prefers OS keychain via `keytar` if available at runtime.
 * Falls back to a local file store with restricted permissions.
 */
export interface ISecureStore {
  get(key: string): Promise<string | null>;
  set(key: string, value: string): Promise<void>;
  del(key: string): Promise<void>;
}

/**
 * Try to lazily import keytar without making it a hard dependency.
 */
async function tryLoadKeytar(): Promise<any | null> {
  try {
    // Avoid static import to prevent type resolution errors when not installed
    const dynamicImport = new Function("m", "return import(m)") as (m: string) => Promise<any>;
    const mod = await dynamicImport("keytar");
    return (mod && (mod.default || mod)) ?? null;
  } catch {
    return null;
  }
}

class KeytarStore implements ISecureStore {
  private service: string;
  private keytar: any;
  constructor(service: string, keytar: any) {
    this.service = service;
    this.keytar = keytar;
  }
  async get(key: string): Promise<string | null> {
    return (await this.keytar.getPassword(this.service, key)) ?? null;
  }
  async set(key: string, value: string): Promise<void> {
    await this.keytar.setPassword(this.service, key, value);
  }
  async del(key: string): Promise<void> {
    await this.keytar.deletePassword(this.service, key);
  }
}

class FileStore implements ISecureStore {
  private filePath: string;

  constructor(namespace: string) {
    const baseDir = FileStore.resolveBaseDir();
    const folder = path.join(baseDir, "uns-auth");
    if (!fs.existsSync(folder)) {
      fs.mkdirSync(folder, { recursive: true, mode: 0o700 });
    }
    this.filePath = path.join(folder, `${FileStore.sanitize(namespace)}.json`);
    if (!fs.existsSync(this.filePath)) {
      fs.writeFileSync(this.filePath, "{}", { mode: 0o600 });
    }
    try {
      // tighten permissions if possible (no-op on Windows)
      fs.chmodSync(this.filePath, 0o600);
    } catch {
      // ignore
    }
  }

  private static resolveBaseDir(): string {
    if (process.platform === "win32") {
      return process.env.APPDATA || path.join(os.homedir(), "AppData", "Roaming");
    }
    // Linux and macOS default to XDG_CONFIG_HOME or ~/.config
    return process.env.XDG_CONFIG_HOME || path.join(os.homedir(), ".config");
  }

  private static sanitize(name: string): string {
    return name.replace(/[^a-zA-Z0-9_.-]/g, "_");
  }

  private async load(): Promise<Record<string, string>> {
    try {
      const raw = await fs.promises.readFile(this.filePath, "utf-8");
      return JSON.parse(raw || "{}");
    } catch {
      return {};
    }
  }

  private async save(data: Record<string, string>): Promise<void> {
    const json = JSON.stringify(data, null, 2);
    await fs.promises.writeFile(this.filePath, json, { mode: 0o600 });
    try {
      fs.chmodSync(this.filePath, 0o600);
    } catch {
      // ignore
    }
  }

  async get(key: string): Promise<string | null> {
    const data = await this.load();
    return data[key] ?? null;
  }

  async set(key: string, value: string): Promise<void> {
    const data = await this.load();
    data[key] = value;
    await this.save(data);
  }

  async del(key: string): Promise<void> {
    const data = await this.load();
    delete data[key];
    await this.save(data);
  }
}

export class SecureStoreFactory {
  /**
   * Creates a secure store for a namespace. Tries keytar first; falls back to file store.
   */
  static async create(namespace: string): Promise<ISecureStore> {
    const keytar = await tryLoadKeytar();
    if (keytar) {
      return new KeytarStore(namespace, keytar);
    }
    return new FileStore(namespace);
  }
}

File: packages/uns-py/src/uns_kit/core/secure_store.py
from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Optional, Protocol


class SecureStore(Protocol):
    def get(self, key: str) -> Optional[str]:
        ...

    def set(self, key: str, value: str) -> None:
        ...

    def delete(self, key: str) -> None:
        ...


class KeyringStore:
    def __init__(self, service: str, keyring_module: object) -> None:
        self.service = service
        self.keyring = keyring_module

    def get(self, key: str) -> Optional[str]:
        value = self.keyring.get_password(self.service, key)
        return value if isinstance(value, str) else None

    def set(self, key: str, value: str) -> None:
        self.keyring.set_password(self.service, key, value)

    def delete(self, key: str) -> None:
        try:
            self.keyring.delete_password(self.service, key)
        except Exception:
            pass


class FileStore:
    def __init__(self, namespace: str, base_dir: Optional[Path] = None) -> None:
        root = (base_dir or self._resolve_base_dir()) / "uns-auth"
        root.mkdir(parents=True, exist_ok=True)
        self.file_path = root / f"{self._sanitize(namespace)}.json"
        if not self.file_path.exists():
            self.file_path.write_text("{}", encoding="utf-8")
        self._chmod(self.file_path, 0o600)

    def get(self, key: str) -> Optional[str]:
        return self._load().get(key)

    def set(self, key: str, value: str) -> None:
        data = self._load()
        data[key] = value
        self._save(data)

    def delete(self, key: str) -> None:
        data = self._load()
        data.pop(key, None)
        self._save(data)

    def _load(self) -> dict[str, str]:
        try:
            raw = self.file_path.read_text(encoding="utf-8")
            parsed = json.loads(raw or "{}")
            return parsed if isinstance(parsed, dict) else {}
        except Exception:
            return {}

    def _save(self, data: dict[str, str]) -> None:
        self.file_path.write_text(json.dumps(data, indent=2), encoding="utf-8")
        self._chmod(self.file_path, 0o600)

    @staticmethod
    def _resolve_base_dir() -> Path:
        if os.name == "nt":
            return Path(os.environ.get("APPDATA", Path.home() / "AppData" / "Roaming"))
        return Path(os.environ.get("XDG_CONFIG_HOME", Path.home() / ".config"))

    @staticmethod
    def _sanitize(name: str) -> str:
        return "".join(char if char.isalnum() or char in "._-" else "_" for char in name)

    @staticmethod
    def _chmod(path: Path, mode: int) -> None:
        try:
            path.chmod(mode)
        except Exception:
            pass


class SecureStoreFactory:
    @staticmethod
    def create(namespace: str, *, base_dir: Optional[Path] = None) -> SecureStore:
        try:
            import keyring  # type: ignore

            return KeyringStore(namespace, keyring)
        except Exception:
            return FileStore(namespace, base_dir=base_dir)
